### DISCUSSION TITLE
Architecture rethinking

### DISCUSSION THREAD

**Discussion by yairyairyair** [OpenFunction/OpenFunction#462] [Ideas]
Hi,

Since i see a lot of parts of open function + a lot of other similar projects or complementary projects i think it will be good to have a discussion on the architecture.

As i see it OpenFunction purpose is to take code (either from container or git) and run it.

- Events: is similar to knative events, however it does have dapr which is a big advantage, seems like be better to implement dapr in knative events.

- functions-framework: The OpenFunction function-framework is google functions-framework but with small changes.

- revision-controller: The revision-controller is minimal and does not support a lot of common use-cases: build triggers instead of polling for example. maybe can use Tekton triggers instead.

- Build: the build part is great in my opinion

- Build Triggers: maybe could add build triggers: like git build trigger, when git repo change or when pull request created.

- Serving: the serving part in my opinion should just be choosing a serving framework (knative/dapr/wasmEdge) and forward the params to the serving framework.

- Installing: the main helm chart is ok but it requires a lot of maintenance, maybe should add an option for user to install the deps by himself and config them as needed and choosing which version that he wants, and OpenFunction should assume these deps CRDs already exists instead of handling deps installation.

In short, maybe we should focus on OpenFunction main features and let the other CRDs via the dependencies do the own features.
This is just my idea, feel free to have discussion.

Thanks
